src/peft/tuners/lora/dora.py

```diff
@@ -86,19 +84,23 @@ def forward(self, x, *, lora_A, lora_B, scaling, base_layer):
         # during backpropagation"
         weight_norm = weight_norm.detach()
         mag_norm_scale = (magnitude / weight_norm).view(1, -1)
-        result_dora = (mag_norm_scale - 1) * (
-            F.linear(x, transpose(weight, self.fan_in_fan_out))
-        ) + mag_norm_scale * lora_result * scaling
-
-        # Note: Computation could potentially be accelerated by using the code below instead of calculating X@W again.
-        # This is only correct if dropout=0, otherwise results will differ:
-        # https://github.com/huggingface/peft/pull/1474#issuecomment-1964682771
-        # bias = self.get_base_layer().bias
-        # if bias is not None:
-        #     result = result - bias
-        # result = mag_norm_scale * result + mag_norm_scale * lora_B(lora_A(x)) * scaling
-        # if bias is not None:
-        #     result = result + bias
+
+        lora_result = lora_B(lora_A(x))
+
+        if base_result is not None:
+            # `base_result` is provided only if dropout is set to 0 or if the model is in evaluation mode.
+            # This means we already have a deterministic output from the base layer that can be reused.
+            bias = base_layer.bias
+            if bias is not None:
+                base_result = base_result - bias
+            result_dora = mag_norm_scale * base_result + mag_norm_scale * lora_result * scaling
```

Comment:
Hmm, wait, should this not be the exact same calculation as in [line 103](https://github.com/huggingface/peft/pull/2122/files#diff-bcb4d7d165949d2eb3eac3203b9067589261e52a1192b7a31f101a3ec98855acR103)? I.e. we should leave the condition after calculating the `base_result` and then do the same calculation of `dora_result` for both cases.

Reply:
I am not sure that I follow. With the `base_result` in place:
1. We first subtract the bias
2. Compute the `dora_result` where the scale the `base_result` with `mag_norm_scale`

But without the `base_result`:
1. We compute the `base_result` with the linear forward
2. Compute the `dora_result` where we scale the `base_result` with `(1 - mag_norm_scale)`

Aren't they going to be different for each case?

Reply:
Okay, so I'm a bit confused, let's try to resolve this.

In the old code, we basically have:

`dora_result = (mag_norm_scale - 1) * base_result + mag_norm_scale * lora_result * lora_scale`

*variable names slightly changed for clarity*

My thinking is that the `base_result` is either calculated right there (old code) or we use the `base_result` that is being passed as an argument, but the basic equation stays the same.

Of course, as you correctly noted, the bias needs to be subtracted first and then added back in the latter case.

In the currently proposed code, in one case we calculate `mag_norm_scale * base_result` and in the other `(mag_norm_scale - 1) * base_result`. This looks inconsistent to me.

Reply:
Hi @BenjaminBossan

I have made the changes as suggested.

Reply:
I think the change is still not correct. Here is my suggestion:

```python
        bias = None
        if base_result is not None:
            bias = base_layer.bias
            if bias is not None:
                result = result - bias
        else:
            base_result = F.linear(x, transpose(weight, self.fan_in_fan_out))

        result_dora = (mag_norm_scale - 1) * base_result + mag_norm_scale * lora_result * scaling

        if bias is not None:
            result_dora = result + bias
```

This way, if `base_result = None`, the computation is exactly the same as it was previously.

I believe the confusion may stem from my comment:

```python
        # result = mag_norm_scale * result + mag_norm_scale * lora_B(lora_A(x)) * scaling
```

This comment should have been:

```python
        # result = (mag_norm_scale - 1) * result + mag_norm_scale * lora_B(lora_A(x)) * scaling
```

Does that make sense?

Reply:
That makes sense!

About the unit test -- do you want me to add a new test file? Or add a test somewhere?